base/iterators.jl

```diff
@@ -1202,7 +1203,23 @@ julia> [(x,y) for x in 0:1 for y in 'a':'c']  # collects generators involving It
 flatten(itr) = Flatten(itr)
 
 eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
-eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}} = promote_typejoin(map(eltype, fieldtypes(I))...)
+
+# For tuples, we statically know the element type of each index, so we can compute
+# this at compile time.
+eltype(::Type{Flatten{I}}) where {I<:Tuple} = _flatten_eltype(Union{}, I)
+
+function eltype(::Type{Flatten{I}}) where {I<:NamedTuple{<:Any, T}} where T
+    _flatten_eltype(Union{}, T)
+end
+
+function _flatten_eltype(T::Type, I::Type{<:Tuple})
+    T2 = promote_typejoin(T, eltype(tuple_type_head(I)))
+    T2 === Any && return Any
+    _flatten_eltype(T2, tuple_type_tail(I))
+end
+
+_flatten_eltype(T::Type, I::Type{Tuple{}}) = T
```

Comment:
If we're willing to give up on the `T2 === Any` short-circuiting, I think we can reduce this to
```suggestion
eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}} =
    mapreduce(eltype, promote_typejoin, fieldtypes(I))
```

Reply:
Good idea - that also constant folds for me. I'm fine with not doing the early return trick.
I'm used your implementation in the latest commit (except added `init=Union{}` to handle the empty named tuple case)